Implement dmx input with rdm support

arneboe opened the pull request:
Currently I am maintaining an old fork of WLED that supports dmx input with rdm using the esp_dmx library.
I would like to get rid of that fork and instead use MoonModules with as little changes as possible.
My usecase is having dmx input with rdm and dmx output work at the same time (using two max495 on port 1 and 2).

I worked through @netmindz dmx input and dmx output branches trying to get them to work and adapt them to my usecase. 

Juggling those branches was kinda challenging for me because they are tracking and old commit of WLED and an old version of esp_dmx and dmx input and output where kinda entangled. Thus I decided to clean it up a bit.
That kinda got out of hand and here we are. (I hope you don't mind :heart: )

The goals for this PR are:
- [x] Have completely different code paths for dmx input and dmx output
- [x] Use esp_dmx for dmx input on esp32 
- [x] Support setting the dmx input address via rdm
- [x] Support setting the dmx input personality via rdm
- [x] Support dmx input rdm identify requests
- [x] Some cleanup and correct error handling

**edit:** From my point of view this PR is ready to merge. It is working as intended in my test cases.

arneboe committed:
Update esp_dmx dependency version to 3.1

arneboe committed:
rename global dmx... variables to dmxInput...

This is the first step in supporting both dmx input and dmx output on different pins.

arneboe committed:
Adapt to new api of esp_dmx v3.1

arneboe committed:
Move dmx_input pin allocations from wled.cpp to dmx.cpp

arneboe committed:
Extract dmx_input from dmx.cpp into dmx_input.cpp.

This greatly improves readability because it gets rid of most of the
ifdefs.

arneboe committed:
Move globals to top of file and change scope to compile unit only.
Some minor cleanup changes

arneboe committed:
Turn dmx_into into class with state.

This is much nicer to read and in the future more state will be added to support all the rdm stuff.

arneboe committed:
handle DMX_INPUT in PinManger::getOwnerText

arneboe committed:
handle dmx rdm identify

netmindz commented:
Sounds like great goals. Let me know if there are any specific questions you have while trying to complete this PR

arneboe committed:
pin esp_dmx to a working commit.

The current release/v3.1 is buggy and cannot receive dmx.
Until that is fixed we are sticking to the last working commit

arneboe commented:
I am having trouble with a crash that happens when dmx is received while the wifi hotspot is being activated.
I am getting the "Cache disabled but cached memory region accessed" panic. 
My guess is, that activating wifi disables the cache.
But the dmx driver has been placed in IRAM so this shouldn't happen? 

When I wait until wifi has been activated before connecting dmx, it works fine. 
But this is not a feasible solution because devices are connected to the dmx bus before they are turned on :D

arneboe commented:
Calling `dmx_receive` with zero timeout does never receive any data.
I am unsure if this is a bug or intended behavior.
I created an issue https://github.com/someweisguy/esp_dmx/issues/87 on esp_dmx to resolve that question.

If this is intended behavior, I will probably have to move the dmx_receive code over into its own task. Do you see any problems with that? And should I pin it to core 0 or 1? Afaik all the arduino stuff runs on core 1, so maybe pinning it to core 0 is a good idea?

arneboe committed:
hack: disable dmx receiver while wifi is being activated

This fixes a crash in the dmx receiver. The dmx receiver cannot work while cache is disabled. For some reason activating wifi disables the cache. In theory, the driver is placed in iram and should work, but it doesn't. This might be a bug in the driver.

arneboe committed:
rename settings

arneboe committed:
add enable/disable methods for dmxInput

arneboe committed:
extract test for rdm identify into own method

arneboe committed:
Monitor dmx personality and dmx start address for change and update rdm

arneboe committed:
extract creation of dmx config into own method

arneboe committed:
handle rdm dmx address changes

arneboe committed:
comments and cleanup

arneboe commented:
> Sounds like great goals. Let me know if there are any specific questions you have while trying to complete this PR

I have a specific question :D
If the user changes the dmx start address using rdm, how can I store it permanently? Probably I have to use the same code path that the web interface is using. But i cannot figure out how the web interface saves data.

arneboe committed:
track v3.1 of esp_dmx

arneboe committed:
track rdm_device_label branch of esp_dmx until PR is done

arneboe committed:
Support dmx rdm personality change

arneboe committed:
keep dmx rdm identify on if dmx disconnects.

Some rdm testers disconnect after setting it.

softhack007 commented:
> [arneboe](https://github.com/arneboe) [force-pushed](https://github.com/MoonModules/WLED/compare/fa4ad138e7f6161982d30c2d2ecb43daf73f7266..5070b1da4a0f617ae7a79e12290430b1bf21b5e2) the dmx_input branch

@arneboe please try to avoid force-pushing. It may cause lots of weird behaviour on github later.

arneboe commented:
> > [arneboe](https://github.com/arneboe) [force-pushed](https://github.com/MoonModules/WLED/compare/fa4ad138e7f6161982d30c2d2ecb43daf73f7266..5070b1da4a0f617ae7a79e12290430b1bf21b5e2) the dmx_input branch
> 
> @arneboe please try to avoid force-pushing. It may cause lots of weird behaviour on github later.
> 
> If you need to copy single commits from one branch to another, it's better to use cherry-picking. It even works between different repositories.
> 
> https://docs.github.com/en/desktop/contributing-and-collaborating-using-github-desktop/managing-commits/cherry-picking-a-commit-in-github-desktop

I am force pushing for several reasons:
- rebasing this branch on top of changes that have happened in mdev since I forked.
- Modifying existing commits because I forgot to push things that belong to that commit.
Both of those shouldn't cause problems when merging later. Rebasing makes the merge easier and modifying my own commits shouldn't matter because they do not exist on the main branch.

I agree, that copying single commits from somewhere should be done by cherry picking. But that is not what I am doing here.

arneboe commented:
The dmx input side is more or less done and I would appreciate some reviews :-)
It is tested and working using a Botex RDM/DMX tester. But that is just one test. Maybe someone else could test it aswell. 

However, it relies on changes that I made to esp_dmx that are stil under review. Thus for now this PR is tracking my fork of esp_dmx until the following PRs are merged:
https://github.com/someweisguy/esp_dmx/pull/90
https://github.com/someweisguy/esp_dmx/pull/92
https://github.com/someweisguy/esp_dmx/pull/93

arneboe committed:
Add dmx input port to configuration

arneboe committed:
Rename WLED_ENABLE_DMX to WLED_ENABLE_DMX_OUTPUT

arneboe committed:
rename dmx.cpp -> dmx_output.cpp

arneboe committed:
rename handleDMX() handleDMXOutput()

arneboe committed:
rename initDmx() -> initDmxOutput()

arneboe commented:
For readability reasons I would like to implement dmx output in a different PR.
Otherwise this gets really hard to review.
Also, I am not sure if there is any benefit in using esp_dmx for the dmx output if the sparkfun lib works just fine? What do you think? @netmindz @softhack007

arneboe committed:
Do no longer disable dmx_input when cache is disabled.

No longer needed because missing ISR_ATTR have been added to esp_dmx.

arneboe committed:
Move dmx_input into its own task on core 0.

This was necessary because otherwise it is not able to respond to rdm in time.

arneboe commented:
I had to put the dmx receiver into its own task on core 0.
Otherwise it was not able to respond to rdm requests in time (there is only a few micros to respond due to how rdm works).
I am not sure if this breaks anything else?.It works for me :D

netmindz commented:
Sorry for not replying, having some health issues. Should hopefully be able to look at soon

arneboe commented:
> Sorry for not replying, having some health issues. Should hopefully be able to look at soon

No need to apoligize :) Take your time!

arneboe committed:
npm run build

arneboe committed:
make compile after rebase

arneboe committed:
npm run build after rebase

arneboe commented:
All required changes have been merged into esp_dmx upstream. Thus I removed my custom branch of esp_dmx. 
I rebased this branch on top of mdev again to keep it easy to merge.

arneboe committed:
switch to esp_dmx v3.1 dependency

netmindz commented on platformio.ini:
```diff
@@ -1038,7 +1038,7 @@ build_flags = ${common.build_flags} ${esp32.build_flagsV4} ${common_mm.build_fla
   -D WLED_ENABLE_DMX_INPUT
 lib_deps = ${esp32.lib_depsV4} ${common_mm.lib_deps_S}
-  https://github.com/someweisguy/esp_dmx.git#v3.0.2-beta  ;; for DMX_INPUT
+  https://github.com/someweisguy/esp_dmx.git#release/v3.1  ;; for DMX_INPUT
```
Please link to a tag rather than branch so we get predictable builds, If no such suitable tag exists, please use known good commit

netmindz commented on tools/cdata.js:
```diff
@@ -359,7 +359,7 @@ writeChunks(
       method: "plaintext",
       filter: "html-minify",
       mangle: (str) => `
-#ifdef WLED_ENABLE_DMX
```
While I understand the reason to rename now we have support for input as well as output, this would be a breaking change and a MR about input doesn't really feel the right place to redefine how to enable output

arneboe replied:
Reverted

netmindz commented on wled00/dmx_input.cpp:
```diff
@@ -0,0 +1,299 @@
+{
+  dmx_config_t config;
+  config.pd_size = 255;
+  config.dmx_start_address = DMXAddress; // TODO split between input and output address
```
Surely output has no start address, only addresses for the fixtures?

netmindz commented on wled00/dmx_input.cpp:
```diff
@@ -0,0 +1,299 @@
+    return;
+  }
+
+  if (rxPin > 0 && enPin > 0 && txPin > 0)
```
Enable is actually optional with the newer RS485 adapters

arneboe replied:
The esp_dmx library requires an en pin. The user can set any unused gpio if their rs485 adapter does not require an en pin. But some pin has to be set.

netmindz replied:
Thought I was passing 0 or -1 before

netmindz replied:
(when not required)

arneboe replied:
I poked around a bit in esp_dmx and it looks like the pin has to fulfill `GPIO_IS_VALID_OUTPUT_GPIO`. Otherwise esp_dmx will fail to initialize.

netmindz commented on wled00/dmx_input.cpp:
```diff
@@ -0,0 +1,299 @@
+    // pin to core 0 because wled is running on core 1
+    xTaskCreatePinnedToCore(dmxReceiverTask, "DMX_RCV_TASK", 10240, this, 2, &task, 0);
+    if (!task)
+    {
```
why the flip in code style to { on new line?

arneboe replied:
vscode auto-formatter. I'all revert that.

netmindz commented on wled00/dmx_input.cpp:
```diff
@@ -0,0 +1,299 @@
+
+void DMXInput::turnOnAllLeds()
+{
+  // TODO not sure if this is the correct way?
```
Would be better to flip to the solid effect with maybe a specific colour?

arneboe replied:
I decided against that because effects might not be available depending on `DmxMode`. Solid is only available if one of the effect modes is selected. For all other modes I would have to enable the leds manually anyway.

netmindz replied:
Fair enough

netmindz commented on wled00/dmx_input.h:
```diff
@@ -0,0 +1,73 @@
+  /// This is the main loop of the dmx receiver. It never returns.
+  friend void dmxReceiverTask(void * context);
+
+  uint8_t inputPortNum = 255; 
```
why initialize as 255?

arneboe replied:
Just some invalid value.

netmindz commented on wled00/dmx_input.h:
```diff
@@ -0,0 +1,73 @@
+  uint8_t enPin = 255;
+
+  /// is written to by the dmx receive task.
+  byte dmxdata[DMX_PACKET_SIZE]; //TODO add locking somehow? maybe double buffer?
```
todo still outstanding?

arneboe replied:
Nope. There is a mutex `dmxDataLock` to guard against that. Just forgot to remove the comment,

netmindz commented on wled00/dmx_output.cpp:
```diff
@@ -1,7 +1,7 @@
 
 /*
- * Support for DMX Output via MAX485.
+ * Support for DMX input and output via serial (e.g. MAX485).
```
Surely your change means it's back to output only?

netmindz commented:
Sorry it's taken so long to get back to you @arneboe

Overall I would say great contribution, you have cleaned up my initial implementation quite a bit, just a few minor issues to address

arneboe commented:
> Sorry it's taken so long to get back to you @arneboe
> 
> Overall I would say great contribution, you have cleaned up my initial implementation quite a bit, just a few minor issues to address

Thanks for the review. I ll fix everything within the next days :)

arneboe committed:
fix: pin esp_dmx version

arneboe committed:
chore: adapt code style

arneboe committed:
chore: remove outdated comments

arneboe committed:
Revert "Rename WLED_ENABLE_DMX to WLED_ENABLE_DMX_OUTPUT"

This reverts commit e3f256bd44981a4beeb14c25391e76977f802184.

netmindz commented:
Sorry, I've not tested this yet @arneboe I've not got a suitable test setup at the moment. I have shared links to this on Discord and asked others to help test but I've not heard back. Can you try and get others to test too? You will also need to resolve the merge conflicts before I can merge

softhack007 commented:
Hi guys @arneboe @netmindz,
I'm preparing for a MM release this weekend. Would it be ok for you if we merge DMX changes after the release?

netmindz commented:
Yeah that's fine @softhack007

netmindz commented:
Are you able to resolve the conflicts @arneboe ?

arneboe commented:
I can try to take a look on the weekend but no promises.

netmindz commented:
Thanks. It's really just the platformio change that matters, for the html stuff it will just be running the npm task and submitting your change

netmindz committed:
Merge branch 'mdev' into dmx_input

netmindz commented:
Github let me resolve the conflict, but then doesn't give me permission to commit the result of "npm run build" @arneboe

netmindz commented:
I've made a few tweaks, if you can merge the the MM dmx_input branch into yours, then I think we are good to then accept this PR

netmindz commented:
You might be interested to know @arneboe this has now been merged upstream into WLED-AV
